### PR TITLE
issue:4030704 Separate the algorithm from the isolation manager to a separate class

### PR DESCRIPTION
**Pull request by vg12345**
## What
Separate the algorithm from the isolation manager to a separate class

## Why ?
Part of separating areas of responsibilities process in the PDR service.
[#4030704](https://redmine.mellanox.com/issues/4030704)
[#4027490](https://redmine.mellanox.com/issues/4027490)

## How ?
The new algorithm class will have two main functions:
    - 1st: receives ports and telemetry data and returns list of detected isolation issues
    - 2nd: checks if given port should be deisolated

## Testing ?
Automatic tests

## Special triggers
_Use the following phrases as comments to trigger different runs_

* ```bot:retest``` rerun Jenkins CI (to rerun GitHub CI, use "Checks" tab on PR page and rerun _all_ jobs)
* ```bot:upgrade``` run additional update tests

### CODE REVIEW

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py [hunk 1]
```diff
@@ -478,15 +478,15 @@ def check_ber_issue(self, port_obj, row, timestamp):
                         return Issue(port_obj.port_name, Constants.ISSUE_BER)
             return None
 
-        def read_next_set_of_high_ber_or_pdr_ports(self):
+        def read_next_set_of_high_ber_or_pdr_ports(self, ports_data, ports_counters):
             """
             Read the next set of ports and check if they have high BER, PDR, temperature or link downed issues
+            Receives: detected ports, isolated ports and telemetry data
+            Returns: lists of isolation issues
```
Comment (drorlevy) [inline plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: isolation candidates

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py [hunk 2]
```diff
@@ -518,7 +518,43 @@ def read_next_set_of_high_ber_or_pdr_ports(self):
                     issues[port_name] = link_downed_issue
                 elif ber_issue:
                     issues[port_name] = ber_issue
-            return issues
+                # If out of operating conditions we'll overwrite the cause
+                if self.temp_check and self.is_out_of_operating_conf(port_name):
+                    issues[port_name] = Constants.ISSUE_OONOC
```
Comment (drorlevy) [inline plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: better renaming "issues" to "isolation_candidates"

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py [hunk 3]
```diff
@@ -478,15 +478,15 @@ def check_ber_issue(self, port_obj, row, timestamp):
                         return Issue(port_obj.port_name, Constants.ISSUE_BER)
             return None
 
-        def read_next_set_of_high_ber_or_pdr_ports(self):
+        def read_next_set_of_high_ber_or_pdr_ports(self, ports_data, ports_counters):
```
Comment (drorlevy) [inline plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: lets rename to something like "analyze_telemetry_data"

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py [hunk 4]
```diff
@@ -841,46 +855,53 @@ def main_flow(self):
...
-                            self.eval_deisolate(port_state.name)
-                ports_updated = self.update_ports_data()
-                if ports_updated:
-                    self.update_telemetry_session()
+                    self.logger.error(f"Failed to read information with error {e}")
+
+                if issues:
+                    if len(issues) > self.max_num_isolate:
+                        # UFM send external event
+                        event_msg = "got too many ports detected as unhealthy: %d, skipping isolation" % len(issues)
```
Comment (drorlevy) [inline plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py:304]: we said this logic isn't the expected behavior. we should add the number of isolated ports in the UFM to the num of ports we want to isolate and then try to isolate as much as we can until we get to max_num_isolate, then issue the warning and skip isolation for the rest

**Review by drorlevy — CHANGES_REQUESTED** [review]

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+            elif link_downed_issue:
+                issues[port_name] = link_downed_issue
+            elif ber_issue:
+                issues[port_name] = ber_issue
+            # If out of operating conditions we'll overwrite the cause
+            if self.temp_check and self.is_out_of_operating_conf(port_name):
+                issues[port_name] = Constants.ISSUE_OONOC
+        return list(issues.values())
+
+    def should_deisolate(self, port_state):
```
Comment (drorlevy) [inline plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py:451]: if this function is doing the port deisolation evaluation and the other function is performing the action then rename it to "eval_port_deisolation" and "perform_deisolation", also add a more informative docstring

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 4 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py:304]: Yes, but is not a part of this PR. You asked not to mix different issues in one PR.
There will be additional PR for this.

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py (hunk 5 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py:451]: This function just checks if given port should be deisolated. It does not perform deisolation,

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 1 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: Ok

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 2 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: This is an old code that just moved. It is called "issues" in all other places. Let's change it later as part of another PR (one of PDR resilience tasks)

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 3 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: Ok

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py (hunk 5 above)
Comment (vg12345) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/pdr_algorithm.py:451]: More explanation added to docstring

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 2 above)
Comment (drorlevy) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: ok just add to the rm board

File: plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py (hunk 4 above)
Comment (drorlevy) [reply plugins/pdr_deterministic_plugin/ufm_sim_web_service/isolation_mgr.py]: ok

**Review by drorlevy — APPROVED** [review]

**Review by TalLerner — APPROVED** [review]

**Pull request merged**